test/jdk/java/util/concurrent/tck/ForkJoinPool20Test.java

```diff
@@ -216,4 +238,416 @@ static ForkJoinWorkerThread submitBusyTask(ForkJoinPool pool) throws Exception {
+     *   cycles - 1 <= elapsedMillis/delay < cycles
+     * Additionally, periodic tasks are not run after shutdown.
+     */
+    public void testFixedRateSequence() throws InterruptedException {
+        final ForkJoinPool p = new ForkJoinPool(4);
+        try (PoolCleaner cleaner = cleaner(p)) {
+            for (int delay = 1; delay <= LONG_DELAY_MS; delay *= 3) {
+                final long startTime = System.nanoTime();
+                final int cycles = 8;
+                final CountDownLatch done = new CountDownLatch(cycles);
+                final Runnable task = new CheckedRunnable() {
+                    public void realRun() { done.countDown(); }};
+                final ScheduledFuture<?> periodicTask =
+                    p.scheduleAtFixedRate(task, 0, delay, MILLISECONDS);
+                final int totalDelayMillis = (cycles - 1) * delay;
+                await(done, totalDelayMillis + LONG_DELAY_MS);
+                final long elapsedMillis = millisElapsedSince(startTime);
+                assertTrue(elapsedMillis >= totalDelayMillis);
+                if (elapsedMillis <= cycles * delay)
+                    return;
+                periodicTask.cancel(true); // retry with longer delay
+            }
+            fail("unexpected execution rate");
+        }
+    }
+    /**
+     * scheduleWithFixedDelay executes series of tasks with given
+     * period.  Eventually, it must hold that each task starts at
+     * least delay and at most 2 * delay after the termination of the
+     * previous task. Additionally, periodic tasks are not run after
+     * shutdown.
+     */
+    public void testFixedDelaySequence() throws InterruptedException {
+        final ForkJoinPool p = new ForkJoinPool(1);
+        try (PoolCleaner cleaner = cleaner(p)) {
+            for (int delay = 1; delay <= LONG_DELAY_MS; delay *= 3) {
+                final long startTime = System.nanoTime();
+                final AtomicLong previous = new AtomicLong(startTime);
+                final AtomicBoolean tryLongerDelay = new AtomicBoolean(false);
+                final int cycles = 8;
+                final CountDownLatch done = new CountDownLatch(cycles);
+                final int d = delay;
+                final Runnable task = new CheckedRunnable() {
+                    public void realRun() {
+                        long now = System.nanoTime();
+                        long elapsedMillis
+                            = NANOSECONDS.toMillis(now - previous.get());
+                        if (done.getCount() == cycles) { // first execution
+                            if (elapsedMillis >= d)
+                                tryLongerDelay.set(true);
+                        } else {
+                            if (elapsedMillis >= 2 * d)
+                                tryLongerDelay.set(true);
+                        }
+                        previous.set(now);
+                        done.countDown();
```

Comment:
Should we test so that we don't count down below 0?

Reply:
Sure. Added
     assertTrue(done.getCount() > 0);
before
                        done.countDown();